src/encoder.rs

```diff
@@ -807,140 +808,580 @@ impl<W: Write> Drop for Writer<W> {
+/// and the CRC will be written at the correct locations in the chunk.
 struct ChunkWriter<'a, W: Write> {
     writer: ChunkOutput<'a, W>,
     buffer: Vec<u8>,
+    /// keeps track of where the last byte was written
     index: usize,
-}
-
-enum ChunkOutput<'a, W: Write> {
-    Borrowed(&'a mut Writer<W>),
-    Owned(Writer<W>),
+    curr_chunk: ChunkType,
 }
 
 impl<'a, W: Write> ChunkWriter<'a, W> {
     fn new(writer: ChunkOutput<'a, W>, buf_len: usize) -> ChunkWriter<'a, W> {
+        // currently buf_len will determine the size of each chunk
+        // the len is capped to the maximum size every chunk can hold
+        // (this wont ever overflow an u32)
+        //
+        // TODO (maybe): find a way to hold two chunks at a time if `usize`
+        //               is 64 bits.
+        const CAP: usize = std::u32::MAX as usize >> 1;
+        let curr_chunk;
+        if writer.sep_def_img || writer.info.frame_control.is_none() || writer.written == 0 {
+            curr_chunk = chunk::IDAT;
+        } else {
+            curr_chunk = chunk::fdAT;
+        }
         ChunkWriter {
             writer,
-            buffer: vec![0; buf_len],
+            buffer: vec![0; CAP.min(buf_len)],
             index: 0,
+            curr_chunk,
         }
     }
-}
 
-impl<'a, W: Write> AsMut<Writer<W>> for ChunkOutput<'a, W> {
-    fn as_mut(&mut self) -> &mut Writer<W> {
-        match self {
-            ChunkOutput::Borrowed(writer) => writer,
-            ChunkOutput::Owned(writer) => writer,
+    /// Returns the size of each scanline for the next frame
+    /// paired with the size of the whole frame
+    ///
+    /// This is used by the `StreamWriter` to know when the scanline ends
+    /// so it can filter compress it and also to know when to start
+    /// the next one
+    fn next_frame_info(&self) -> (usize, usize) {
+        let wrt = self.writer.deref();
+
+        let width: usize;
+        let height: usize;
+        if let Some(fctl) = wrt.info.frame_control {
+            width = fctl.width as usize;
+            height = fctl.height as usize;
+        } else {
+            width = wrt.info.width as usize;
+            height = wrt.info.height as usize;
         }
+
+        let in_len = wrt.info.raw_row_length_from_width(width as u32) - 1;
+        let data_size = in_len * height;
+
+        (in_len, data_size)
     }
-}
 
-impl<W: Write> Write for ChunkWriter<'_, W> {
-    fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
-        let written = buf.read(&mut self.buffer[self.index..])?;
-        self.index += written;
+    /// NOTE: this bypasses the internal buffer so the flush method should be called before this
+    ///       in the case there is some data left in the buffer when this is called, it will panic
+    fn write_header(&mut self) -> Result<()> {
+        if self.index != 0 {
+            // return Err(todo!())
+            panic!("Called when not flushed") // TODO: add error code
```

Comment:
TODO?

Reply:
it was more like a question if an error code should be added, as this code is only called internally and it can fail only if the implementation is wrong

Reply:
If the implementation requires this internally then you can use `assert_eq!(self.index, 0)`.